File: python/pytest-reporter/src/llm_pytest_reporter/error_classifier.py
"""Error classification and analysis utilities."""

import re
from typing import List, Dict, Optional, Tuple
from .models import ErrorInfo, TestResult, TestSuite


class ErrorClassifier:
    """Classifies and analyzes test errors."""
    
    # Common error patterns
    ASSERTION_PATTERNS = [
        (r"AssertionError", "Assertion Error"),
        (r"assert.*==", "Equality Assertion"),
        (r"assert.*!=", "Inequality Assertion"),
        (r"assert.*is", "Identity Assertion"),
        (r"assert.*in", "Membership Assertion"),
        (r"assertTrue|assertFalse", "Boolean Assertion"),
    ]
    
    EXCEPTION_PATTERNS = [
        (r"TypeError", "Type Error"),
        (r"ValueError", "Value Error"),
        (r"KeyError", "Key Error"),
        (r"IndexError", "Index Error"),
        (r"AttributeError", "Attribute Error"),
        (r"NameError", "Name Error"),
        (r"ImportError|ModuleNotFoundError", "Import Error"),
        (r"ZeroDivisionError", "Division by Zero"),
        (r"FileNotFoundError", "File Not Found"),
        (r"PermissionError", "Permission Error"),
        (r"TimeoutError", "Timeout Error"),
    ]
    
    ASYNC_PATTERNS = [
        (r"asyncio.*TimeoutError", "Async Timeout"),
        (r"RuntimeError.*await", "Async Runtime Error"),
        (r"not awaited", "Missing Await"),
    ]
    
    def classify_error(self, error: ErrorInfo) -> str:
        """Classify an error into a category."""
        error_text = f"{error.type} {error.message}"
        
        # Check assertion patterns
        for pattern, category in self.ASSERTION_PATTERNS:
            if re.search(pattern, error_text, re.IGNORECASE):
                return category
        
        # Check exception patterns
        for pattern, category in self.EXCEPTION_PATTERNS:
            if re.search(pattern, error_text, re.IGNORECASE):
                return category
        
        # Check async patterns
        for pattern, category in self.ASYNC_PATTERNS:
            if re.search(pattern, error_text, re.IGNORECASE):
                return category
        
        # Default classification based on error type
        if "Error" in error.type:
            return error.type
        
        return "Unknown Error"
    
    def generate_fix_hint(self, error: ErrorInfo) -> str:
        """Generate a fix hint based on error type and content."""
        error_type = error.type.lower()
        message = error.message.lower()
        
        # Type errors
        if "typeerror" in error_type:
            if "got" in message and "expected" in message:
                return "Check the types of arguments being passed"
            elif "not callable" in message:
                return "Ensure you're calling a function, not a value"
            elif "missing" in message and "argument" in message:
                return "Check function signature for required arguments"
            return "Verify data types match expected values"
        
        # Value errors
        elif "valueerror" in error_type:
            if "invalid literal" in message:
                return "Check input format and conversion logic"
            return "Ensure values are within expected range or format"
        
        # Key/Index errors
        elif "keyerror" in error_type:
            return "Verify dictionary key exists before access"
        elif "indexerror" in error_type:
            return "Check array bounds before accessing elements"
        
        # Attribute errors
        elif "attributeerror" in error_type:
            if "has no attribute" in message:
                return "Verify object has the attribute or method"
            return "Check object type and available attributes"
        
        # Import errors
        elif "importerror" in error_type or "modulenotfounderror" in error_type:
            return "Ensure module is installed and import path is correct"
        
        # Assertion errors
        elif "assertionerror" in error_type:
            if error.expected and error.actual:
                return f"Update test or implementation to match expected value"
            return "Review assertion logic and expected values"
        
        # Timeout errors
        elif "timeout" in error_type or "timeout" in message:
            return "Increase timeout duration or optimize async operations"
        
        # File errors
        elif "filenotfound" in error_type:
            return "Verify file path and ensure file exists"
        
        # Permission errors
        elif "permission" in error_type:
            return "Check file/directory permissions"
        
        # Default hint
        return "Review error message and stack trace for details"
    
    def extract_values(self, message: str) -> Tuple[Optional[str], Optional[str]]:
        """Extract expected and actual values from error message."""
        # Common assertion patterns
        patterns = [
            # pytest style: assert x == y
            r"assert\s+(.+?)\s*==\s*(.+)",
            # unittest style: Expected: x, but was: y
            r"Expected:\s*(.+?),\s*but\s*was:\s*(.+)",
            # Jest style: Expected x to equal y
            r"Expected\s+(.+?)\s+to\s+(?:equal|be)\s+(.+)",
            # Custom: expected x, got y
            r"expected\s+(.+?),\s*got\s+(.+)",
        ]
        
        for pattern in patterns:
            match = re.search(pattern, message, re.IGNORECASE)
            if match:
                actual = match.group(1).strip()
                expected = match.group(2).strip()
                return expected, actual
        
        return None, None
    
    def detect_patterns(self, suites: List[TestSuite]) -> Dict[str, List[TestResult]]:
        """Detect patterns across multiple test failures."""
        patterns: Dict[str, List[TestResult]] = {}
        
        for suite in suites:
            for test in suite.tests:
                if test.error:
                    category = self.classify_error(test.error)
                    if category not in patterns:
                        patterns[category] = []
                    patterns[category].append(test)
        
        # Filter out single occurrences
        return {k: v for k, v in patterns.items() if len(v) > 1}
    
    def generate_pattern_summary(self, patterns: Dict[str, List[TestResult]]) -> str:
        """Generate a summary of detected patterns."""
        if not patterns:
            return ""
        
        output = "\n## DETECTED PATTERNS\n"
        for category, tests in sorted(patterns.items(), key=lambda x: len(x[1]), reverse=True):
            output += f"- {category}: {len(tests)} occurrences\n"
            # Show first few test names
            for test in tests[:3]:
                output += f"  - {test.full_name}\n"
            if len(tests) > 3:
                output += f"  ... and {len(tests) - 3} more\n"
        
        return output

File: go/shared/error_classifier.go
package shared

import (
	"regexp"
	"strings"
)

// ErrorClassifier classifies and enhances error information
type ErrorClassifier struct {
	patterns map[string]*regexp.Regexp
}

// NewErrorClassifier creates a new error classifier
func NewErrorClassifier() *ErrorClassifier {
	return &ErrorClassifier{
		patterns: map[string]*regexp.Regexp{
			"expected_got":     regexp.MustCompile(`[Ee]xpected\s+(.+?)\s+but\s+got\s+(.+)`),
			"expected_actual":  regexp.MustCompile(`expected:\s*(.+?)\s*actual:\s*(.+)`),
			"want_got":        regexp.MustCompile(`want\s+(.+?),\s*got\s+(.+)`),
			"should_be":       regexp.MustCompile(`(.+?)\s+should\s+be\s+(.+)`),
			"not_equal":       regexp.MustCompile(`(.+?)\s*!=\s*(.+)`),
			"assertion":       regexp.MustCompile(`assertion failed`),
			"nil_pointer":     regexp.MustCompile(`nil pointer|null pointer|nil reference`),
			"index_bounds":    regexp.MustCompile(`index out of (range|bounds)|array index`),
			"type_mismatch":   regexp.MustCompile(`type mismatch|cannot convert|incompatible type`),
			"timeout":         regexp.MustCompile(`timeout|timed out|deadline exceeded`),
			"panic":          regexp.MustCompile(`panic:|runtime error:`),
		},
	}
}

// ClassifyError determines the type of error
func (ec *ErrorClassifier) ClassifyError(message string) string {
	lowerMsg := strings.ToLower(message)
	
	// Check specific patterns
	if ec.patterns["nil_pointer"].MatchString(lowerMsg) {
		return "Nil Pointer"
	}
	if ec.patterns["index_bounds"].MatchString(lowerMsg) {
		return "Index Error"
	}
	if ec.patterns["type_mismatch"].MatchString(lowerMsg) {
		return "Type Error"
	}
	if ec.patterns["timeout"].MatchString(lowerMsg) {
		return "Timeout"
	}
	if ec.patterns["panic"].MatchString(message) {
		return "Panic"
	}
	if ec.patterns["assertion"].MatchString(lowerMsg) || 
	   ec.patterns["expected_got"].MatchString(message) ||
	   ec.patterns["not_equal"].MatchString(message) {
		return "Assertion Error"
	}
	
	return "Error"
}

// ExtractValues extracts expected and actual values from error message
func (ec *ErrorClassifier) ExtractValues(message string) (expected, actual string) {
	// Try different patterns
	if matches := ec.patterns["expected_got"].FindStringSubmatch(message); len(matches) > 2 {
		return strings.TrimSpace(matches[1]), strings.TrimSpace(matches[2])
	}
	if matches := ec.patterns["expected_actual"].FindStringSubmatch(message); len(matches) > 2 {
		return strings.TrimSpace(matches[1]), strings.TrimSpace(matches[2])
	}
	if matches := ec.patterns["want_got"].FindStringSubmatch(message); len(matches) > 2 {
		return strings.TrimSpace(matches[1]), strings.TrimSpace(matches[2])
	}
	if matches := ec.patterns["should_be"].FindStringSubmatch(message); len(matches) > 2 {
		return strings.TrimSpace(matches[2]), strings.TrimSpace(matches[1])
	}
	
	return "", ""
}

// GenerateFixHint generates a hint for fixing the error
func (ec *ErrorClassifier) GenerateFixHint(errorInfo *ErrorInfo) string {
	errorType := errorInfo.Type
	message := strings.ToLower(errorInfo.Message)
	
	switch errorType {
	case "Nil Pointer":
		return "Check for nil values before dereferencing"
	case "Index Error":
		return "Verify array/slice bounds before accessing"
	case "Type Error":
		return "Ensure type compatibility in operations"
	case "Timeout":
		return "Increase timeout duration or optimize the operation"
	case "Panic":
		return "Add error handling or recover from panic"
	case "Assertion Error":
		if errorInfo.Expected != "" && errorInfo.Actual != "" {
			return "Review assertion logic and expected values"
		}
		return "Check test assertions and expectations"
	}
	
	// Generic hints based on message content
	if strings.Contains(message, "not found") || strings.Contains(message, "does not exist") {
		return "Verify resource exists and path is correct"
	}
	if strings.Contains(message, "permission") || strings.Contains(message, "access denied") {
		return "Check file permissions and access rights"
	}
	if strings.Contains(message, "connection") || strings.Contains(message, "network") {
		return "Verify network connectivity and endpoints"
	}
	
	return "Review error message and stack trace for details"
}